### PR TITLE
vmware: add the vmware_guest_instantclone module

### PR DESCRIPTION
**Pull request by PDQDakota**
##### SUMMARY
Adds support for the instant clone API in vSphere to the vmware_guest module.

Instant Cloning is a method of cloning a running or frozen VM that uses Copy on Write for both disk and RAM allocation in order to reduce the amount of resources used by a cloned VM and to reduce the time required for the clone to become available. It requires a vSphere instance for the API to be available, it doesn't work with only an ESXi host.

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
vmware_guest module

##### ADDITIONAL INFORMATION
I don't think I have more to add here but please request anything further you'd need of me!

### CODE REVIEW

**Comment by pgbidkar** [comment]
My suggestion on this PR would be, change looks big,  create separate module something like vmware_guest_instant_clone, instead of adding it to vmware_guest. 
vmware_guest is already over loaded and vulnerable to breakages. 

Thanks

**Comment by PDQDakota** [comment]
@pgbidkar I'll see about moving it out into a separate module then.

**Comment by goneri** [comment]
Hi @PDQDakota, thanks for the contribution and welcome in the community! We try to cover all the new features with little test playbooks.
See: https://docs.ansible.com/ansible/devel/dev_guide/platforms/vmware_guidelines.html

Do you think you can extend the existing functional tests to cover this new use-case? This way will be able to react faster in the future if something breaks your code. This will also speed up the review process.  For the record, the other tests are located in this directory: https://github.com/ansible/ansible/tree/devel/test/integration/targets/vmware_guest/tasks

**Comment by PDQDakota** [comment]
I'm getting close to adding the requested changes but I'm hitting a snag with my integration tests.
According to this guide [here](https://docs.ansible.com/ansible/devel/dev_guide/platforms/vmware_guidelines.html#configure-your-installation) the presence of a `cloud-config-vcenter.ini` copied and edited from the `cloud-config-vcenter.ini.template` file should ensure that it's possible to run vmware integration tests.
However, when I do run my integration test or any other vmware integration test they are skipped.
The output being:
```
WARNING: Excluding tests marked "cloud/vcenter" which require config (see "/home/dakota/ansible/test/lib/ansible_test/config/cloud-config-vcenter.ini.template"): vmware_guest_instantclone
WARNING: All targets skipped.
```
However, I have followed the instructions correctly. Would anyone know what I am doing wrong?
I can paste the contents of my ini file (sans sensitive info) if that would help!

**Comment by pgbidkar** [comment]
Hi @PDQDakota , Can you check this 
```
ERROR: lib/ansible/module_utils/vmware.py:327:4: bare-except: No exception type(s) specified (100%)
``` 
There is no Exception mentioned here
```
try:
        facts['instant_clone_frozen'] = vm.summary.runtime.instantCloneFrozen
    except:
        # This property is only available to vSphere 6.7 and later
        # This value will only affect instant clone operations
        # which are not available to all ESXi and vSphere installations.
        pass
```
this is not allowed

**Comment by pgbidkar** [comment]
Integration test https://github.com/ansible/ansible/pull/61014/files#diff-0593167dedf1ed2881e8691f6714e481 format doesn't look correct

**Comment by pgbidkar** [comment]
Here
```
- All parameters and VMware object names are case sensitive.
version_added: 2.9
```
Change to
```
version_added: 2.10
```

**Comment by PDQDakota** [comment]
@pgbidkar I"ve made the changes to the version and updated the exception pass as well.

As for the integration test, I can't seem to get to the point where it will run at all, it's just skipped.
I'll make the format corrections, I'd just like to get to the point it tells me something more than that it was skipped, even when run directly.
I've copied and update the ini file as instructed in the template ini. The documentation page for the integration doesn't seem to mention any further required steps.
Can you help me figure out what I'm doing wrong? Or is there a better place to ask for that assistent?

File: lib/ansible/modules/cloud/vmware/vmware_guest_instantclone.py [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+
+            for vm_network_adapter in vm_network_adapters:
+                # Standard network switch
+                if isinstance(vm_network_adapter.backing, vim.VirtualEthernetCardNetworkBackingInfo):
+                    network_id = vm_network_adapter.backing.network
+                    device_spec = vim.VirtualDeviceConfigSpec()
+                    device_spec.operation = 'edit'
+                    device_spec.device = vm_network_adapter
+                    device_spec.device.backing = vim.VirtualEthernetCardNetworkBackingInfo()
+                    device_spec.device.backing.deviceName = network_id
```
Comment (goneri) [inline lib/ansible/modules/cloud/vmware/vmware_guest_instantclone.py:167]: `network_id` is a `vim.Network` instance, so you should actually set `network_id.name` in `deviceName`.

File: test/integration/targets/vmware_guest_instantclone/tasks/main.yml [hunk 2]
```diff
@@ -0,0 +1,9 @@
+# Test0001: Create an instant clone
+- name: 0001 - Create an instant clone
+    validate_certs: False
+    hostname: '{{ vcenter_hostname }}'
+    username: '{{ vcenter_username }}'
+    password: '{{ vcenter_password }}'
+    datacenter: '{{ dc1 }}'
+    name: '{{ source_vm }}'
+    clone_name: '{{ name }}'
```
Comment (goneri) [inline test/integration/targets/vmware_guest_instantclone/tasks/main.yml:9]: You need to initialize the environment with `prepare_vmware_test`. I used this playbook to test your module:
```yaml
- import_role:
    name: prepare_vmware_tests
  vars:
    setup_attach_host: true
    setup_datastore: true
    setup_virtualmachines: true

- name: set state to poweron the first VM
  vmware_guest_powerstate:
    validate_certs: no
    hostname: "{{ vcenter_hostname }}"
    username: "{{ vcenter_username }}"
    password: "{{ vcenter_password }}"
    name: "{{ virtual_machines[0].name }}"
    folder: '{{ f0 }}'
    state: powered-on

# Test0001: Create an instant clone
- name: 0001 - Create an instant clone
  vmware_guest_instantclone:
    validate_certs: False
    hostname: '{{ vcenter_hostname }}'
    username: '{{ vcenter_username }}'
    password: '{{ vcenter_password }}'
    datacenter: '{{ dc1 }}'
    name: "{{ virtual_machines[0].name }}"
    clone_name: test_vm1
```

**Review by goneri — CHANGES_REQUESTED** [review]

**Review by goneri — CHANGES_REQUESTED** [review]
Nice module :-). Just a little comment, otherwise everything looks good to me.

Could you also add the follwing content in test/integration/targets/vmware_guest_instantclone/aliases:
```
cloud/vcenter
shippable/vcenter/group1
needs/target/prepare_vmware_tests
```

**Comment by Akasurde** [comment]
Thank you very much for your interest in Ansible. This plugin is no longer maintained in this repository and has been migrated to https://github.com/ansible-collections/community.vmware.

Please re-submit this issue in the above repository.

If you have further questions please stop by IRC or the mailing list:

* IRC: #ansible on irc.freenode.net
* mailing list: https://groups.google.com/forum/#!forum/ansible-project

**Pull request closed without merging**